code/modules/integrated_electronics/subtypes/manipulation.dm

```diff
@@ -203,4 +203,177 @@
+	activators = list("pulse in" = IC_PINTYPE_PULSE_IN,"pulse out" = IC_PINTYPE_PULSE_OUT)
+	spawn_flags = IC_SPAWN_RESEARCH
+	power_draw_per_use = 50
+/*
+/obj/item/integrated_circuit/manipulation/thrower/New()
+	processing_objects |= src
+/obj/item/integrated_circuit/manipulation/thrower/Destroy()
+	processing_objects -= src
+/obj/item/integrated_circuit/manipulation/thrower/process()
+	set_pin_data(IC_OUTPUT, 1, weakref(contents[1]))
+	set_pin_data(IC_OUTPUT, 2, weakref(contents[contents.len]))
+	set_pin_data(IC_OUTPUT, 3, src.contents.len)
+	push_data()
+*/
+/obj/item/integrated_circuit/manipulation/grabber/do_work()
+	var/turf/T = get_turf(src)
+	var/atom/movable/AM = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	var/mode = get_pin_data(IC_INPUT, 2)
+	if(mode == 1)
+		if(AM.Adjacent(T))
+			if(contents.len < 10)
+				if(istype(AM,/obj/item))
+					var obj/item/A = AM
+					if(A.w_class < ITEMSIZE_NORMAL)
+						AM.loc = src
+				else if(istype(AM,/mob))
+					var mob/A = AM
+					if(istiny(A))
+						AM.loc = src
+	if(mode == 0)
+		if(contents.len)
+			var/obj/item/U = contents[1]
+			U.loc = T
+	if(mode == -1)
+		if(contents.len)
+			var/obj/item/U
+			for(U in contents)
+				U.loc = T
+	set_pin_data(IC_OUTPUT, 1, weakref(contents[1]))
+	set_pin_data(IC_OUTPUT, 2, weakref(contents[contents.len]))
+	set_pin_data(IC_OUTPUT, 3, src.contents.len)
+	push_data()
+	activate_pin(2)
+
+
+
+/obj/item/integrated_circuit/manipulation/thrower
+	name = "thrower"
+	desc = "A compact launcher to throw things from inside or nearby tiles"
+	extended_desc = "The first and second inputs need to be numbers.  They are coordinates to throw thing at, relative to the machine itself.  \
+	The 'fire' activator will cause the mechanism to attempt to throw thing at the coordinates, if possible.  Note that the \
+	projectile need to be inside the machine, or to be on an adjacent tile, and to be up to medium size."
+	complexity = 15
+	w_class = ITEMSIZE_NORMAL
+	size = 2
+	inputs = list(
+		"target X rel" = IC_PINTYPE_NUMBER,
+		"target Y rel" = IC_PINTYPE_NUMBER,
+		"projectile" = IC_PINTYPE_REF
+		)
+	outputs = list()
+	activators = list(
+		"fire" = IC_PINTYPE_PULSE_IN
+	)
+	spawn_flags = IC_SPAWN_RESEARCH
+	origin_tech = list(TECH_ENGINEERING = 3, TECH_DATA = 3, TECH_COMBAT = 4)
+	power_draw_per_use = 50
+
+
+
+/obj/item/integrated_circuit/manipulation/thrower/do_work()
+
+	var/datum/integrated_io/target_x = inputs[1]
+	var/datum/integrated_io/target_y = inputs[2]
+	var/datum/integrated_io/projectile = inputs[3]
+	if(!isweakref(projectile.data))
+		return
+	var/obj/item/A = projectile.data.resolve()
+	if(A.anchored)
+		return
+	if(A.w_class>ITEMSIZE_NORMAL)
+		return
+	var/turf/T = get_turf(src.assembly)
+	var/turf/TP = get_turf(A)
+	if(!(TP.Adjacent(T)))
+		return
+	if(src.assembly)
+		if(isnum(target_x.data))
+			target_x.data = round(target_x.data)
+		if(isnum(target_y.data))
+			target_y.data = round(target_y.data)
+
+
```

Comment:
```suggestion
```
Unneeded whitespace

Reply:
these things came with the code I lifted it from. All I know is they work.